template/web/app/paymaster-bundler/_utils/fetchNFTs.ts

```diff
@@ -0,0 +1,60 @@
+import { SmartAccountClient } from 'permissionless';
+import { PublicClient } from 'viem';
+
+import { nftAbi } from '../_components/abi';
+import { NFTType } from '../types';
+
+export default async function fetchNFTs(smartAccount: SmartAccountClient, client: PublicClient) {
+  // Get # of NFTs owned by address
+  const address = smartAccount.account?.address;
+
+  const numTokens = await client.readContract({
+    address: '0x66519FCAee1Ed65bc9e0aCc25cCD900668D3eD49',
+    abi: nftAbi,
+    functionName: 'balanceOf',
+    args: [address],
+  });
+
+  // Get the token IDs and metadata by token ID
+  var tokens = [] as string[];
+
+  for (let i = 0; i < Number(numTokens); i++) {
+    const tokenID = await client.readContract({
+      address: '0x66519FCAee1Ed65bc9e0aCc25cCD900668D3eD49',
+      abi: nftAbi,
+      functionName: 'tokenOfOwnerByIndex',
+      args: [address, i],
+    });
+
+    const tokenJSONLink = await client.readContract({
+      address: '0x66519FCAee1Ed65bc9e0aCc25cCD900668D3eD49',
+      abi: nftAbi,
+      functionName: 'tokenURI',
+      args: [Number(tokenID)],
+    });
+
+    tokens.push(tokenJSONLink as string);
+  }
+
+  const fetchOps = [] as Promise<unknown>[];
+  const tokenJSONs = [] as NFTType[];
+
+  tokens.forEach((token) => {
```

Comment:
This is crazy inefficient for large collections, it will end up locking up the UI and sending a large amount of requests to the RPC provider.  I know this is not a simple solution, however, we should probably use a data aggregator here. cc: @Zizzamia .. WDYT about integrating with an API vendor for these type of aggregated data lookups?